trie/structural.go

```diff
@@ -120,3 +133,430 @@ func generateStructLen(buffer []byte, l int) int {
+		s.itemTypes = append(s.itemTypes, AccountStreamItem)
+	case hashNode:
+		if trace {
+			fmt.Printf("hashNode %x\n", hex)
+		}
+		hashOnly := rs.HashOnly(hex)
+		if !hashOnly {
+			if c := rs.Current(); len(c) == len(hex)+1 && c[len(c)-1] == 16 {
+				hashOnly = true
+			}
+		}
+		if hashOnly {
+			var hn common.Hash
+			copy(hn[:], []byte(n))
+			s.hexes = append(s.hexes, hex)
+			s.hashes = append(s.hashes, hn)
+			if accounts {
+				s.itemTypes = append(s.itemTypes, AHashStreamItem)
+			} else {
+				s.itemTypes = append(s.itemTypes, SHashStreamItem)
+			}
+		} else {
+			panic(fmt.Errorf("unexpected hashNode: %s, at hex: %x (%d)", n, hex, len(hex)))
+		}
+	default:
+		panic(fmt.Errorf("unexpected node: %T", nd))
+	}
+}
+
+// StreamHash computes the hash of a stream, as if it was a trie
+func StreamHash(s *Stream, trace bool) (common.Hash, error) {
+	hb := NewHashBuilder(trace)
+	var succ bytes.Buffer
+	var curr OneBytesTape
+	var value OneBytesTape
+	var hashes TwoHashTape
+	var groups, sGroups []uint16 // Separate groups slices for storage items and for accounts
+	var a accounts.Account
+	var fieldSet uint32
+	var ki, ai, si, hi int
+	var itemType StreamItem
+	hb.SetKeyTape(&curr)
+	hb.SetValueTape(&value)
+	hb.SetNonceTape((*OneUint64Tape)(&a.Nonce))
+	hb.SetBalanceTape((*OneBalanceTape)(&a.Balance))
+	hb.SetHashTape(&hashes)
+	hb.SetSSizeTape((*OneUint64Tape)(&a.StorageSize))
+	hb.Reset()
+	curr.Reset()
+	hashOnly := func(_ []byte) bool { return true }
+	for ki < len(s.hexes) {
+		hex := s.hexes[ki]
+		curr.Reset()
+		curr.Write(succ.Bytes())
+		succ.Reset()
+		succ.Write(hex)
+		if curr.Len() > 0 {
+			var err error
+			if itemType == AccountStreamItem || itemType == AHashStreamItem {
+				groups, err = GenStructStep(fieldSet, hashOnly, itemType == AHashStreamItem, false, curr.Bytes(), succ.Bytes(), hb, groups)
+			} else {
+				sGroups, err = GenStructStep(fieldSet, hashOnly, itemType == SHashStreamItem, false, curr.Bytes(), succ.Bytes(), hb, sGroups)
+			}
+			if err != nil {
+				return common.Hash{}, err
+			}
+		}
+		itemType = s.itemTypes[ki]
+		ki++
+		switch itemType {
+		case AccountStreamItem:
+			a.Copy(s.aValues[ai])
+			ai++
+			if a.IsEmptyCodeHash() && a.IsEmptyRoot() {
+				fieldSet = AccountFieldSetNotContract
+			} else {
+				if a.HasStorageSize {
+					fieldSet = AccountFieldSetContractWithSize
+				} else {
+					fieldSet = AccountFieldSetContract
+				}
+				// Load hashes onto the stack of the hashbuilder
+				hashes.hashes[0] = a.CodeHash // this will be just beneath the top of the stack
+				//hashes.hashes[1] = a.Root     // this will end up on top of the stack
+				hashes.idx = 0 // Reset the counter
+				// the first item ends up deepest on the stack, the seccond item - on the top
+				if err := hb.hash(1); err != nil {
+					return common.Hash{}, err
+				}
+			}
+		case StorageStreamItem:
+			v := s.sValues[si]
+			si++
+			value.Buffer.Reset()
+			if len(v) > 1 || v[0] >= 128 {
```

Comment:
could you extract it to a separate method?
Or add a comment/const for 128.

Reply:
This will go away after https://github.com/ledgerwatch/turbo-geth/pull/170 is merged, and I will wait for 170 to be merged first :)